File: app/src/main/java/com/incite/moviequiz/ArcadeFilmsViewActivity.java
package com.incite.moviequiz;

import android.content.Intent;

import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Point;
import android.os.Bundle;
import android.view.Display;
import android.view.View;
import android.view.Window;
import android.widget.GridLayout;
import android.widget.ImageView;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.core.app.ActivityOptionsCompat;
import androidx.core.content.ContextCompat;
import androidx.core.view.ViewCompat;
import android.transition.Fade;

import com.airbnb.lottie.LottieAnimationView;
import com.bumptech.glide.Glide;
import com.google.android.material.imageview.ShapeableImageView;
import com.google.android.material.shape.CornerFamily;
import com.incite.moviequiz.data.Data;
import com.incite.moviequiz.data.Film;

import java.util.ArrayList;

//770 X 450 IMG
public class ArcadeFilmsViewActivity extends AppCompatActivity {

    LottieAnimationView sound;
    TextView levelTv;
    TextView moneyTv;
    ArrayList<ShapeableImageView> images;

    int id;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_arcadefilmsview);

        Fade fade = new Fade();
        View decor = getWindow().getDecorView();
        fade.excludeTarget(android.R.id.statusBarBackground, true);
        fade.excludeTarget(android.R.id.navigationBarBackground, true);
        fade.excludeTarget(R.id.rlfv, true);
        fade.excludeTarget(R.id.rlag, true);
        getWindow().setEnterTransition(fade);
        getWindow().setExitTransition(fade);

        sound = findViewById(R.id.soundOnOff2);
        sound.setMinFrame(60);
        if(!SoundManager.isOn) sound.setFrame(60);
        else sound.setFrame(90);

        levelTv = findViewById(R.id.allFilmsTv2);
        moneyTv = findViewById(R.id.moneyTv2);
        images = new ArrayList<>();


        Intent intent = getIntent();
        id = intent.getIntExtra("packId", 0);
        Player.isLastMain = true;
        Player.currentPackId = id;

        /** SET PASSLEVEL TV **/
        int passed = 0;
        int all = 0;
        for (int j = 0; j < Data.getPacks().get(id).size(); j++) {
            if (Data.getPacks().get(id).getFilms().get(j).isPassed()) {
                passed++;
            }
            all++;
        }
        String str = passed + " / " + all;
        levelTv.setText(str);
        moneyTv.setText(String.valueOf(Player.money));
        //


        Display display = getWindowManager().getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);

        ScrollView scrollView = findViewById(R.id.mainScrollView);


        GridLayout gridLayout = new GridLayout(this);
        gridLayout.setColumnCount(2);



        int pixelsMarginSides = (int) (0.05 * size.x);
        int pixelsMarginTop = (int) (0.05 * size.x);
        int imageWidth = (int) (0.4 * size.x);
        int imageHeight = (int) (0.234 * size.x);


        for (int j = 0; j < Data.getPacks().get(id).size(); j++) {
            Film currentFilm = Data.getPacks().get(id).getFilms().get(j);



            GridLayout.LayoutParams imageViewLp = new GridLayout.LayoutParams();

            imageViewLp.rowSpec = GridLayout.spec(0, 1f);
            imageViewLp.width = imageWidth;
            imageViewLp.height = imageHeight;



            ShapeableImageView imageView = new ShapeableImageView(this);
            imageView.setTransitionName("cardViewTransition");
            imageView.setId(j);
            imageView.setContentDescription(currentFilm.getAnswer());

            imageViewLp.setMargins(pixelsMarginSides, pixelsMarginTop - 12, pixelsMarginSides, 0);
            imageView.setLayoutParams(imageViewLp);


            //imageView.setImageDrawable(ContextCompat.getDrawable(this,currentFilm.getDrawableID()));
            Glide.with(getApplicationContext()).load(currentFilm.getDrawableID()).into(imageView);



            imageView.setShapeAppearanceModel(imageView.getShapeAppearanceModel()
                    .toBuilder()
                    .setAllCorners(CornerFamily.ROUNDED,getResources().getDimension(R.dimen.default_corner_radius))
                    .build());


            if (Data.getPacks().get(id).getFilms().get(j).isPassed()) {
                ColorMatrix matrix = new ColorMatrix();
                matrix.setSaturation(0);
                ColorMatrixColorFilter filter = new ColorMatrixColorFilter(matrix);
                imageView.setColorFilter(filter);
            }

            imageView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Intent i = new Intent(ArcadeFilmsViewActivity.this, ArcadeGameActivity.class);
                    i.putExtra("filmId", (view.getId()));
                    i.putExtra("packId", Player.currentPackId);
                    Player.currentFilmId = (view.getId());

                    ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(ArcadeFilmsViewActivity.this, imageView, ViewCompat.getTransitionName(imageView));
                    i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                    SoundManager.play(SoundManager.Click1);
                    startActivity(i, options.toBundle());
                    //supportFinishAfterTransition();
                }
            });
            gridLayout.addView(imageView);
            images.add(imageView);

        }
        scrollView.addView(gridLayout);
    }

    public static ColorMatrixColorFilter brightIt(int fb) {
        ColorMatrix cmB = new ColorMatrix();
        cmB.set(new float[]{
                1, 0, 0, 0, fb,
                0, 1, 0, 0, fb,
                0, 0, 1, 0, fb,
                0, 0, 0, 1, 0});

        ColorMatrix colorMatrix = new ColorMatrix();
        colorMatrix.set(cmB);

        ColorMatrixColorFilter f = new ColorMatrixColorFilter(colorMatrix);
        return f;
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        Intent i = new Intent(ArcadeFilmsViewActivity.this, ArcadeActivity.class);
        startActivity(i);
        finish();
    }

    @Override
    protected void onRestart() {
        super.onRestart();

        Player.isLastMain = true;
        for (int i = 0; i < images.size(); i++) {
            if (Data.getPacks().get(id).getFilms().get(i).isPassed()) {
                ColorMatrix matrix = new ColorMatrix();
                matrix.setSaturation(0);
                ColorMatrixColorFilter filter = new ColorMatrixColorFilter(matrix);
                images.get(i).setColorFilter(filter);
            }
        }
        int passed = 0;
        int all = 0;
        for (int j = 0; j < Data.getPacks().get(id).size(); j++) {
            if (Data.getPacks().get(id).getFilms().get(j).isPassed()) {
                passed++;
            }
            all++;
        }
        String str = passed + " / " + all;
        levelTv.setText(str);
        moneyTv.setText(String.valueOf(Player.money));
        //
    }

    public void startShopActivity(View view) {
        Intent i = new Intent(ArcadeFilmsViewActivity.this, ShopActivity.class);
        startActivity(i);
        finish();
    }

    public void checkSound(View view) {
        if (SoundManager.isOn) {
            SoundManager.play(SoundManager.Bubble);
            SoundManager.isOn = false;
            SoundManager.stop(SoundManager.Tick);
            sound.setSpeed(-1.5f);
        } else {
            SoundManager.isOn = true;
            SoundManager.play(SoundManager.Bubble);
            sound.setSpeed(1.5f);
        }
        sound.playAnimation();
    }
}

File: app/src/main/java/com/incite/moviequiz/presentation/arcade_films/ArcadeFilmsViewActivity.kt
package com.incite.moviequiz.presentation.arcade_films

import com.incite.moviequiz.util.SoundManager.isOn
import com.incite.moviequiz.util.SoundManager.play
import com.incite.moviequiz.util.SoundManager.Click1
import com.airbnb.lottie.LottieAnimationView
import android.widget.TextView
import com.google.android.material.imageview.ShapeableImageView
import android.os.Bundle
import com.incite.moviequiz.R
import android.content.Intent
import android.widget.ScrollView
import com.bumptech.glide.Glide
import com.google.android.material.shape.CornerFamily
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Point
import android.net.Uri
import android.transition.Fade
import android.view.View
import android.widget.GridLayout
import androidx.appcompat.app.AppCompatActivity
import com.incite.moviequiz.presentation.arcade_game.ArcadeGameActivity
import androidx.core.app.ActivityOptionsCompat
import androidx.core.view.ViewCompat
import com.incite.moviequiz.domain.model.DataLoader
import com.incite.moviequiz.domain.model.Player
import com.incite.moviequiz.presentation.arcade.ArcadeActivity
import com.incite.moviequiz.presentation.shop.ShopActivity
import com.incite.moviequiz.util.SoundManager
import dagger.hilt.android.AndroidEntryPoint
import java.util.ArrayList
import javax.inject.Inject

@AndroidEntryPoint
//770 X 450 IMG
class ArcadeFilmsViewActivity : AppCompatActivity() {

    @Inject
    lateinit var dataLoader: DataLoader
    private var player: Player? = null

    private lateinit var sound: LottieAnimationView
    private lateinit var levelTv: TextView
    private lateinit var moneyTv: TextView
    private lateinit var images: ArrayList<ShapeableImageView>
    private var id = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_arcadefilmsview)

        setAnimations()
        initViews()
        setViews()

        player = dataLoader.repo.getPlayer()

        if (!isOn) sound.frame = 60 else sound.frame = 90

        id = intent.getIntExtra("packId", 0)

        /** SET PASSLEVEL TV  */


        var passed = 0
        var all = 0
        for (j in 0 until dataLoader.packs[id].films.size) {
            if (dataLoader.packs[id].films[j].completed) {
                passed++
            }
            all++
        }
        val str = "$passed / $all"
        levelTv.text = str
        moneyTv.text = player?.money.toString()

        val display = windowManager.defaultDisplay
        val size = Point()
        display.getSize(size)
        val scrollView = findViewById<ScrollView>(R.id.mainScrollView)
        val gridLayout = GridLayout(this)
        gridLayout.columnCount = 2
        val pixelsMarginSides = (0.05 * size.x).toInt()
        val pixelsMarginTop = (0.05 * size.x).toInt()
        val imageWidth = (0.4 * size.x).toInt()
        val imageHeight = (0.234 * size.x).toInt()
        for (j in 0 until dataLoader.packs[id].films.size) {
            val currentFilm = dataLoader.packs[id].films[j]
            val imageViewLp = GridLayout.LayoutParams()
            imageViewLp.rowSpec = GridLayout.spec(0, 1f)
            imageViewLp.width = imageWidth
            imageViewLp.height = imageHeight
            val imageView = ShapeableImageView(this)
            imageView.transitionName = "cardViewTransition"
            imageView.id = j
            imageView.contentDescription = currentFilm.name
            imageViewLp.setMargins(pixelsMarginSides, pixelsMarginTop - 12, pixelsMarginSides, 0)
            imageView.layoutParams = imageViewLp

            Glide.with(applicationContext).load(Uri.parse(currentFilm.image)).into(imageView)
            imageView.shapeAppearanceModel = imageView.shapeAppearanceModel
                .toBuilder()
                .setAllCorners(
                    CornerFamily.ROUNDED,
                    resources.getDimension(R.dimen.default_corner_radius)
                )
                .build()
            if (dataLoader.packs[id].films[j].completed) {
                val matrix = ColorMatrix()
                matrix.setSaturation(0f)
                val filter = ColorMatrixColorFilter(matrix)
                imageView.colorFilter = filter
            }
            imageView.setOnClickListener { view ->
                val i = Intent(this@ArcadeFilmsViewActivity, ArcadeGameActivity::class.java)
                i.putExtra("filmId", view.id)
                i.putExtra("packId", id)
                i.putExtra("isLastMain", true)
                i.putExtra("isLastPrev", false)
                val options = ActivityOptionsCompat.makeSceneTransitionAnimation(
                    this@ArcadeFilmsViewActivity,
                    imageView,
                    ViewCompat.getTransitionName(imageView)!!
                )
                i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                play(Click1)
                startActivity(i, options.toBundle())
            }
            gridLayout.addView(imageView)
            images.add(imageView)
        }
        scrollView.addView(gridLayout)
    }

    private fun setAnimations(){
        val fade = Fade()
        fade.excludeTarget(android.R.id.statusBarBackground, true)
        fade.excludeTarget(android.R.id.navigationBarBackground, true)
        fade.excludeTarget(R.id.rlfv, true)
        fade.excludeTarget(R.id.rlag, true)
        window.enterTransition = fade
        window.exitTransition = fade
    }

    private fun initViews(){
        sound = findViewById(R.id.soundOnOff2)
        levelTv = findViewById(R.id.allFilmsTv2)
        moneyTv = findViewById(R.id.moneyTv2)

    }

    private fun setViews(){
        sound.setMinFrame(60)
        images = ArrayList()

    }


    override fun onBackPressed() {
        super.onBackPressed()
        val i = Intent(this@ArcadeFilmsViewActivity, ArcadeActivity::class.java)
        startActivity(i)
        finish()
    }

    override fun onRestart() {
        super.onRestart()
        for (i in images.indices) {
            if (dataLoader.packs[id].films[i].completed) {
                val matrix = ColorMatrix()
                matrix.setSaturation(0f)
                val filter = ColorMatrixColorFilter(matrix)
                images[i].colorFilter = filter
            }
        }
        var passed = 0
        var all = 0
        for (j in 0 until dataLoader.packs[id].films.size) {
            if (dataLoader.packs[id].films[j].completed) {
                passed++
            }
            all++
        }
        val str = "$passed / $all"
        levelTv.text = str
        moneyTv.text = player?.money.toString()
    }

    fun startShopActivity(view: View?) {
        val i = Intent(this@ArcadeFilmsViewActivity, ShopActivity::class.java)
        startActivity(i)
        finish()
    }

    fun checkSound(view: View?) {
        if (SoundManager.isOn) {
            SoundManager.play(SoundManager.Bubble);
            SoundManager.isOn = false;
            SoundManager.stop(SoundManager.Tick);
            sound.speed = -1.5f
        } else {
            SoundManager.isOn = true;
            SoundManager.play(SoundManager.Bubble);
            sound.speed = 1.5f
        }
        sound.playAnimation()
    }

}